### ISSUE TITLE
Memory Leak? 

### ISSUE THREAD

**Issue by PaulIPS** [evermeer/EVReflection#60]
Hi I am using version 2.20.2. There appears to be a memory leak when converting to and from JSON, more so when converting from JSON to an object. 

My JSON has more values and levels but I was able to reproduce it with a simpiler structure. 

``` json
{
    "name" : "Test",
    "listOfObjects" : [
        {
            "name" : "Object 1",
            "value" : "Value 1"   
        }
    ]
}
```

Create about 8000 of the list objects, and try and convert to JSON. You will see a spike of memory. Try to convert the JSON back to objects and you will see a big spike in memory. 

If you profile in instruments you will see the JSON lines of code come up with memory leak warning. A very quick look at the allocations and it seams like there are a few items in side of EVReflect that are probably causing most of the issues. I haven't gotten a chance to dig in to the code and see if I can figure out where exactly it's coming from. If I get a chance to look into it more I will let you know.

**Comment by evermeer** [comment]
Thanks for reporting this. I will look into it later this weekend. Of course pointers to the location of the problem are welcome.

**Comment by evermeer** [comment]
In the latest version all Swift dictionaries and arrays where converted to NSDictionary and NSArray. This resulted in a performance improvement of about 15% but then I also added an autoreleaspool block on 2 crucial locations. That one had a negative impact of about 15% but the memory leak was less than half. I will do some tests using instruments (with and without the autoreleasepool's) this weekend to see if I can find the leek.

**Comment by evermeer** [comment]
I added a test app so that the instruments app can be used to track memory usage and leaks. 
I have not used Instruments often. I do see memory leeks, but I don't know how I can track where these come from. Does anyone else know how I can find the source of the leak?

**Comment by evermeer** [comment]
I have found a memory leak in  the following locations:

In a NSJSONSerialisation function. It's from Apple, how can I solve this?

In the EVReflection valueForAny function (a dynamic cast, but where?) I will investigate this

In the following 2 simple functions. What's wrong with it?

```
    public class func swiftClassFromString(className: String) -> NSObject? {
        return (swiftClassTypeFromString(className) as? NSObject.Type)?.init()
    }
    public class func swiftStringFromClass(theObject: NSObject) -> String! {
        return NSStringFromClass(theObject.dynamicType).stringByReplacingOccurrencesOfString(getCleanAppName(theObject) + ".", withString: "", options: NSStringCompareOptions.CaseInsensitiveSearch, range: nil)
    }

```

**Comment by JRG-Developer** [comment]
> I have found a memory leak in the following locations:
> 
> In a NSJSONSerialisation function. It's from Apple, how can I solve this?

I'm fairly sure this is a bug in `NSJSONSerialisation` at least on certain versions of iOS, which makes it difficult/impossible to fix.

How do you feel about using an open-source parser instead of `NSJSONSerialisation`, such as [SBJson](https://github.com/stig/json-framework)?

Reasoning:  Apple is notoriously slow for fixing bugs, and even when they do, they'll still be around for _years_ due to the need for apps to support prior iOS major versions.

By adopting an open-source solution, this gives developers full control of which library version is used, and they can also contribute to bug fixes themselves, which can be used _immediately_.

**Comment by evermeer** [comment]
I think I would rather add [Freddy](https://github.com/bignerdranch/Freddy)
But I do have too manny other things to do so it could take some time before I have time to pick something like this up

**Comment by evermeer** [comment]
It looks like I have solved the memory leak. See the details below.

From the Apple Swift release documentation for the 7.1 release at:
https://developer.apple.com/library/ios/releasenotes/DeveloperTools/RN-Xcode/Chapters/xc7_release_notes.html

> Using switch against multiple types with as patterns no longer causes a memory leak. (22587077)

But apparently it did not solve the issue for me.

The leak was in this part of the code:

```
        switch theValue {
        // Bool, Int, UInt, Float and Double are casted to NSNumber by default!
        case let numValue as NSNumber:
            return (numValue, "NSNumber", false)
        case let longValue as Int64:
            return (NSNumber(longLong: longValue), "NSNumber", false)
        case let longValue as UInt64:
            return (NSNumber(unsignedLongLong: longValue), "NSNumber", false)
        case let intValue as Int32:
            return (NSNumber(int: intValue), "NSNumber", false)
        case let intValue as UInt32:
            return (NSNumber(unsignedInt: intValue), "NSNumber", false)
        case let intValue as Int16:
            return (NSNumber(short: intValue), "NSNumber", false)
        case let intValue as UInt16:
            return (NSNumber(unsignedShort: intValue), "NSNumber", false)
        case let intValue as Int8:
            return (NSNumber(char: intValue), "NSNumber", false)
        case let intValue as UInt8:
            return (NSNumber(unsignedChar: intValue), "NSNumber", false)
        case let stringValue as NSString:
            return (stringValue, "NSString", false)
        case let stringValue as String:
            return (NSString(string: stringValue), "NSString", false)
        case let dateValue as NSDate:
            return (dateValue, "NSDate", false)
        case let anyvalue as NSArray:
            return (anyvalue, valueType, false)
        case let anyvalue as EVObject:
            if valueType.containsString("<") {
                return (anyvalue, swiftStringFromClass(anyvalue), true)
            }
            return (anyvalue, valueType, true)
        case let anyvalue as NSObject:
            if valueType.containsString("<") {
                return (anyvalue, swiftStringFromClass(anyvalue), false)
            }
            // isObject is false to prevent parsing of objects like CKRecord, CKRecordId and other objects.
            return (anyvalue, valueType, false)
        default:
            (parentObject as? EVObject)?.addStatusMessage(.InvalidType, message: "valueForAny unkown type \(valueType) for value: \(theValue).")
            print("ERROR: valueForAny unkown type \(valueType) for value: \(theValue).")
            return (NSNull(), "NSNull", false)
        }
```

After converting it to this the leak was gone:

```
        if theValue is NSNumber {
            return (theValue as! NSNumber, "NSNumber", false)
        }
        if theValue is Int64 {
            return (NSNumber(longLong: theValue as! Int64), "NSNumber", false)
        }
        if theValue is UInt64 {
            return (NSNumber(unsignedLongLong: theValue as! UInt64), "NSNumber", false)
        }
        if theValue is Int32 {
            return (NSNumber(int: theValue as! Int32), "NSNumber", false)
        }
        if theValue is UInt32 {
            return (NSNumber(unsignedInt: theValue as! UInt32), "NSNumber", false)
        }
        if theValue is Int16 {
            return (NSNumber(short: theValue as! Int16), "NSNumber", false)
        }
        if theValue is UInt16 {
            return (NSNumber(unsignedShort: theValue as! UInt16), "NSNumber", false)
        }
        if theValue is Int8 {
            return (NSNumber(char: theValue as! Int8), "NSNumber", false)
        }
        if theValue is UInt8 {
            return (NSNumber(unsignedChar: theValue as! UInt8), "NSNumber", false)
        }
        if theValue is NSString {
            return (theValue as! NSString, "NSString", false)
        }
        if theValue is NSDate {
            return (theValue as! NSDate, "NSDate", false)
        }
        if theValue is NSArray {
            return (theValue as! NSArray, valueType, false)
        }
        if theValue is EVObject {
            if valueType.containsString("<") {
                return (theValue as! EVObject, swiftStringFromClass(theValue as! EVObject), true)
            }
            return (theValue as! EVObject, valueType, true)
        }
        if theValue is NSObject {
            if valueType.containsString("<") {
                return (theValue as! NSObject, swiftStringFromClass(theValue as! NSObject), true)
            }
            return (theValue as! NSObject, valueType, true)
        }
        (parentObject as? EVObject)?.addStatusMessage(.InvalidType, message: "valueForAny unkown type \(valueType) for value: \(theValue).")
        print("ERROR: valueForAny unkown type \(valueType) for value: \(theValue).")
        return (NSNull(), "NSNull", false)
```

The memory leak was reported like this:

<img width="1012" alt="screen shot 2016-05-19 at 9 52 27 pm" src="https://cloud.githubusercontent.com/assets/130834/15408690/bed36ee2-1e11-11e6-841a-78103ac20292.png">

I will upload a new version with the fix in a moment.

**Comment by evermeer** [comment]
I just pushed this fix as version 2.33.0 to GitHub and Cocoapods

**Comment by evermeer** [comment]
The memory leak in the switch case is a know issue. For more information see https://bugs.swift.org/browse/SR-1339?jql=text%20~%20%22switch%20case%20leak%22

**Comment by PaulIPS** [comment]
@evermeer I finally got a chance to try the newest version of the library against my test case and it performs much better. Memory is greatly improved, and the speed of json to object is about 3 times faster. 

I do think there is still a small memory issue going on. I have noticed that if I do the same to json or from json multiple times each time memory increases slightly. The bigger the objects the more it builds up.
